kore/src/Kore/Strategies/Goal.hs

```diff
@@ -618,40 +679,108 @@ checkImplication' lensRulePattern goal =
+            unificationResults <-
+                OrPattern.observeAllT unificationProblems
+            when
+                (all isBottom unificationResults)
+                (succeed . NotImplied $ claimPattern)
+            removals <-
+                OrPattern.observeAllT (removedDestinations unificationResults)
+            when (all isBottom removals) (succeed Implied)
+            simplifiedRemovals <- simplifyConjunctionOfRemovals removals
+            when (isBottom simplifiedRemovals) (succeed Implied)
+            let stuckConfiguration = OrPattern.toPattern simplifiedRemovals
+            claimPattern
+                & Lens.set (field @"left") stuckConfiguration
                 & NotImpliedStuck
                 & pure
         & run
+      | otherwise =
+      error . show . Pretty.vsep $
+          [ "The check implication step expects\
+          \ the configuration term to be function-like."
+          , Pretty.indent 2 "Configuration term:"
+          , Pretty.indent 4 (unparse leftTerm)
+          ]
       where
-        configuration = Lens.view RulePattern.leftPattern rulePattern
-        configFreeVars = freeVariables configuration
-        destination =
-            Lens.view (field @"rhs") rulePattern
-            & topExistsToImplicitForall configFreeVars
-        existentials =
-            Lens.view (field @"existentials")
-            . Lens.view (field @"rhs")
-            $ rulePattern
+        ClaimPattern { right, left, existentials } = claimPattern
+        leftFreeVars = ClaimPattern.freeVariablesLeft claimPattern
+        right' =
+            ClaimPattern.topExistsToImplicitForall leftFreeVars existentials
+            <$> right
+        leftTerm = Pattern.term left
+        leftCondition = Pattern.withoutTerm left
+        leftSort = termLikeSort leftTerm
+        sideCondition =
+            SideCondition.assumeTrueCondition
+                (Condition.fromPredicate . Condition.toPredicate $ leftCondition)
+
+        unificationProblems
+            :: LogicT
+                (ExceptT (CheckImplicationResult ClaimPattern) m)
+                (OrPattern RewritingVariableName)
+        unificationProblems = do
+            rightPatt <- Logic.scatter right'
+            let rightTerm = Pattern.term rightPatt
+            mkIn leftSort leftTerm rightTerm
+                & Pattern.fromTermLike
+                & Pattern.simplify sideCondition
+
+        removedDestinations
+            :: MultiOr (OrPattern RewritingVariableName)
+            -> LogicT
+                (ExceptT (CheckImplicationResult ClaimPattern) m)
+                (OrPattern RewritingVariableName)
+        removedDestinations unificationResults = do
+            singleUnificationCondition <-
+                Logic.scatter
+                    $ (fmap . fmap) Conditional.withoutTerm unificationResults
+            rightCondition <-
+                Logic.scatter (Conditional.withoutTerm <$> right')
```

Comment:
The second part of the problem is that this line mixes up which `ensures` goes with which unification problem. I suggest that we keep each `ensures` with its corresponding term. Using `OrPattern` throughout, instead of mixing `LogicT` and `OrPattern`, would help with that.